## Task 1: Dataset Understanding

In [82]:
import pandas as pd

# Load dataset

In [83]:
df = pd.read_csv("customer_support_text_classification.csv")


# 1. Number of records

In [84]:
print("Number of records:", len(df))


Number of records: 1500


# 2. Target labels/classes

In [85]:
print("\nTarget labels:")
print(df["sentiment_label"].unique())


Target labels:
<StringArray>
['neutral', 'positive', 'negative']
Length: 3, dtype: str


# 3. Sample text records

In [86]:
print("\nSample customer messages:")
print(df["customer_message"].head())



Sample customer messages:
0    I need information about the payment process. ...
1        I need information about the payment process.
2    The refund process was fast and convenient. I ...
3    My refund is still pending and this experience...
4     Please tell me how to update my account details.
Name: customer_message, dtype: str


# 4. Average text length

In [87]:
text_lengths = df["customer_message"].apply(lambda x: len(x.split()))

average_length = text_lengths.mean()

print("\nAverage text length:", average_length)


Average text length: 12.722666666666667


# 5. Class distribution

In [88]:
print("\nClass distribution:")
print(df["sentiment_label"].value_counts())


Class distribution:
sentiment_label
neutral     524
negative    497
positive    479
Name: count, dtype: int64


## Task 2: Text Preprocessing


In [89]:


import pandas as pd
import re

df = pd.read_csv("customer_support_text_classification.csv")

stop_words = ["the", "is", "and", "a", "an", "to", "of", "in", "for", "on", "with", "was", "my"]

def clean_text(text):
    text = text.lower()
    text = re.sub(r'[^a-zA-Z\s]', '', text)
    words = text.split()
    words = [word for word in words if word not in stop_words]
    return words

df["cleaned_text"] = df["customer_message"].apply(clean_text)

print(df[["customer_message", "cleaned_text"]].head())

                                    customer_message  \
0  I need information about the payment process. ...   
1      I need information about the payment process.   
2  The refund process was fast and convenient. I ...   
3  My refund is still pending and this experience...   
4   Please tell me how to update my account details.   

                                        cleaned_text  
0  [i, need, information, about, payment, process...  
1    [i, need, information, about, payment, process]  
2  [refund, process, fast, convenient, i, appreci...  
3  [refund, still, pending, this, experience, fru...  
4  [please, tell, me, how, update, account, details]  


## Task 3 Text Vectorization

In [90]:


import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer


df = pd.read_csv("customer_support_text_classification.csv")


vectorizer = CountVectorizer()


X = vectorizer.fit_transform(df["customer_message"])


print("Shape of vector data:", X.shape)


print(X.toarray()[:5])

Shape of vector data: (1500, 667)
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


Why text is converted into vectors:

Machine learning models cannot understand text directly.
They only work with numbers.

Text vectorization converts words into numerical form so the model can learn patterns from the text data.

Observation:
The text data was converted into numerical vectors using Bag of Words.
The dataset contains 1500 messages and 667 unique words/features.

## Task 4 Baseline Model

In [91]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score

# Load dataset

In [92]:
df = pd.read_csv("customer_support_text_classification.csv")

# Convert text into numbers

In [93]:
vectorizer = CountVectorizer()

X = vectorizer.fit_transform(df["customer_message"])


# Target column

In [94]:
y = df["sentiment_label"]

# Split data

In [95]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)


# Create model


In [96]:
model = MultinomialNB()

# Train model


In [97]:
model.fit(X_train, y_train)
predictions = model.predict(X_test)

# Check accuracy


In [98]:
accuracy = accuracy_score(y_test, predictions)
print("Model Accuracy:", accuracy)

Model Accuracy: 1.0


Observation:
The Naive Bayes model achieved very high accuracy on the dataset.
This means the model was able to correctly classify customer sentiments using text vectors.

## Task 5: Sequence Model Architecture

Model Selected: LSTM

1. Input Sequence:
The customer message is converted into a sequence of numbers using a tokenizer.

2. Embedding Layer:
The embedding layer converts words into dense vectors.

3. LSTM Layer:
The LSTM layer processes the sequence and remembers important information from previous words.

4. Output Layer:
The output layer predicts the sentiment class:
positive, negative, or neutral.

5. Loss Function:
Categorical Crossentropy

6. Evaluation Metric:
Accuracy

Input Text
→ Tokenizer
→ Padding
→ Embedding Layer
→ LSTM Layer
→ Dense Output Layer
→ Sentiment Prediction

## Task 6: Attention and Transformer Reflection

1. Why RNNs struggle with long-term dependencies:
RNN models read text one word at a time. When the sentence becomes very long, the model may forget earlier information. Because of this, RNNs sometimes cannot understand long sentences properly.

2. How LSTMs help with memory:
LSTMs are improved versions of RNNs. They contain memory cells that help the model remember important information for a longer time. This helps in understanding long sentences better.

3. What attention solves in sequence-to-sequence tasks:
The attention mechanism helps the model focus on important words in a sentence. Instead of using only the final output, the model can look at all important words while making predictions.

4. Why transformers are important in modern NLP and Generative AI:
Transformers are faster and more powerful than RNNs and LSTMs. They can process many words at the same time and understand relationships between words better. Modern AI tools like ChatGPT and Google Gemini use transformer models.